File: elixir/minesweeper/lib/minesweeper.ex
defmodule Matrix do
  defstruct [:grid, :r, :c]

  def parse(input) do
    grid = input
    |> Enum.map(&String.graphemes/1)
    |> Enum.map(&List.to_tuple/1)
    |> List.to_tuple()
    r = grid |> tuple_size()
    c = if r == 0, do: 0, else: grid |> elem(0) |> tuple_size()
    %Matrix{grid: grid, r: r, c: c}
  end

  def get(matrix, {i, j}), do: matrix.grid |> elem(i) |> elem(j)

  def adjacent(matrix, {i, j}) do
    for x <- i - 1 .. i + 1,
        y <- j - 1 .. j + 1,
        x != i or y != j,
        0 <= x and x < matrix.r,
        0 <= y and y < matrix.c, do: {x, y}
  end

end
defmodule Minesweeper do
  @doc """
  Annotate empty spots next to mines with the number of mines next to them.
  """
  @spec annotate([String.t()]) :: [String.t()]

  def annotate(board) do
    board
    |> Matrix.parse()
    |> build_list_matrix()
    |> Enum.map(&Enum.join(&1, ""))
  end

  defp build_list_matrix(%Matrix{grid: _, r: 0, c: 0}), do: []
  defp build_list_matrix(%Matrix{grid: _, r: r, c: 0}), do: Stream.repeatedly(fn -> [] end) |> Enum.take(r)
  defp build_list_matrix(matrix) do
    for i <- 0 .. matrix.r - 1,
        j <- 0 .. matrix.c - 1 do
          {i, j}
        end
    |> Enum.map(&get_value(matrix, &1))
    |> Enum.chunk_every(matrix.c)
  end

  defp get_value(matrix, position) do
    case Matrix.get(matrix, position) do
      " " ->
        count = matrix
        |> Matrix.adjacent(position)
        |> Enum.filter(&(Matrix.get(matrix, &1) == "*"))
        |> Enum.count()
        case count do
          0 -> " "
          _ -> count |> Integer.to_string()
        end
      value -> value
    end
  end

end

File: haskell/minesweeper/src/Minesweeper.hs
module Minesweeper (annotate) where
import Data.List (groupBy)

annotate :: [String] -> [String]
annotate [""] = [""]
annotate board = (unindexBoard . parse . indexBoard) board

indexBoard :: [String] -> [(Int, Int, Char)]
indexBoard = concatMap (\(x, row) -> [(x, y, c) | (y, c) <- zip [0..] row])
    . zip [0..] 

unindexBoard :: [(Int, Int, Char)] -> [String]
unindexBoard = (map . map) (\(_, _, c) -> c)
    . groupBy (\(x1, _, _) (x2, _, _) -> x1 == x2)

adjacentIndexes :: Int -> Int -> [(Int, Int, Char)] -> [(Int, Int, Char)]
adjacentIndexes x y = filter (\(i, j, _) -> i /= x || j /= y)
    . filter (\(i, _, _) -> abs(i - x) <= 1)
    . filter (\(_, j, _) -> abs(j - y) <= 1)

countBombs :: Int -> Int -> [(Int, Int, Char)] -> Int
countBombs x y = length
    . filter (\(_, _, c) -> c == '*')
    . adjacentIndexes x y

mapValue :: (Int, Int, Char) -> [(Int, Int, Char)] -> (Int, Int, Char)
mapValue (x, y, '*') _ =  (x, y, '*')
mapValue (x, y, _) board = (x, y, c)
    where
        count = countBombs x y board
        c = if count == 0 then ' ' else (head . show) count

parse :: [(Int, Int, Char)] -> [(Int, Int, Char)]
parse board = map (`mapValue` board) board

